From 3323b9ffb8ea3d7026a3709b99d8393547e3e77c Mon Sep 17 00:00:00 2001
From: Arnon Kaufman <akaufman@habana.ai>
Date: Sun, 22 Oct 2023 18:15:40 +0300
Subject: [PATCH] [SW-15774] build: refactor zephyr package and scripts

As part of separating the targets build procedures,
a fundamental change in the related files have to be do.

Previous commits lay down the basics make files to create the different
targets although those new make files are not automatically triggered.
Due to the nature of the buildroot concepts, the products will be
created in different location.
In order to keep backward compatibility, the zephyr target should be
kept.
The need to keep the zephyr build and deal with the inherit change of
directory, the zephyr makefile and the support scripts should be changed
as part of the same commit.

For backward compatibility the zephyr target is kept. Consolidate the
individual packages, as list of dependencies. Calling the zephyr target
will trigger the dependencies, making the targets
zephyr-{dirclean,rebuild,...} available.

The additional new targets:
pid, mgmt, event, preboot,
agent-fwl, agent-serdes, agent-pcie, agent-hbm

Example:
make pid-dirclean
make pid-rebuild

make zephyr-dirclean # trigger all depended -dirclean rule
make zephyr-rebuild  # trigger all dependencies

Change-Id: I5a9dca37a65ec686bde42ce80cff55457df4e6c7
Signed-off-by: Arnon Kaufman <akaufman@habana.ai>
---

diff --git a/board/habanalabs/common/post-image-agents.sh b/board/habanalabs/common/post-image-agents.sh
index c256dfd..61ad2e8 100755
--- a/board/habanalabs/common/post-image-agents.sh
+++ b/board/habanalabs/common/post-image-agents.sh
@@ -21,7 +21,7 @@
 # Set hl-*-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>]"; inside FIT header
 # Set hl-preboot-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>] (<date> - <time>)"
 HL_PREBOOT_VERSION=`cat ${BASE_DIR}/build/baremetal-custom/buildversion-habanalabs`
-HL_AGENT_VERSION=`cat ${BASE_DIR}/build/zephyr-custom/build-agent-fw-loader/buildversion-habanalabs`
+HL_AGENT_VERSION=`cat ${BASE_DIR}/build/agent-fwl-custom/build/buildversion-habanalabs`
 # Set enum number of FIT_LOAD_TYPE_FLASH in FIT
 FIT_LOAD_TYPE_FLASH=$(grep -E "LOAD_TYPE_FLASH[^_]" ${SPECS_EMBEDDED_DIR}/fit_load_type.h | sed 's/.* = \(.*\),/\1/')
 FIT_LOAD_TYPE_RUN=$(grep -E "FIT_LOAD_TYPE_RUN[^_]" ${SPECS_EMBEDDED_DIR}/fit_load_type.h | sed 's/.* = \(.*\),/\1/')
diff --git a/board/habanalabs/common/post-image-fit-global.sh b/board/habanalabs/common/post-image-fit-global.sh
index 7fa938d..aa84e94 100755
--- a/board/habanalabs/common/post-image-fit-global.sh
+++ b/board/habanalabs/common/post-image-fit-global.sh
@@ -34,8 +34,8 @@
 HL_SEC_KEY_INDEX=`grep HL_SEC_KEY_INDEX $HL_SEC_FILE | sed 's/^.*\(0x.*\)$/\1/'`
 HL_SEC_FW_VERSION=`grep HL_SEC_FW_VERSION $HL_SEC_FILE | awk '{ print $NF }'`
 # set zephyr apps version
-HL_ZEPHYR_MGMT_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-mgmt/buildversion-habanalabs`
-HL_ZEPHYR_PID_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-pid/buildversion-habanalabs`
+HL_MGMT_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/mgmt-custom/build/buildversion-habanalabs`
+HL_PID_VERSION=`cat ${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/pid-custom/build/buildversion-habanalabs`
 
 COMMON_BASE=${BR2_EXTERNAL}/board/habanalabs/common
 UBOOT_ZEPHYR_ITS_FILE=uboot-zephyr-${DEV_NAME}_${HL_TARGET}.its
@@ -123,8 +123,8 @@
 ITS_SED_PARAMS+="s/@PPBOOT_SEC_OFFSET@/${PPBOOT_SEC_OFFSET}/g;"
 
 #add zephyr app versions
-ITS_SED_PARAMS+="s/@HL_ZEPHYR_MGMT_VERSION@/${HL_ZEPHYR_MGMT_VERSION}/g;"
-ITS_SED_PARAMS+="s/@HL_ZEPHYR_PID_VERSION@/${HL_ZEPHYR_PID_VERSION}/g;"
+ITS_SED_PARAMS+="s/@HL_MGMT_VERSION@/${HL_MGMT_VERSION}/g;"
+ITS_SED_PARAMS+="s/@HL_PID_VERSION@/${HL_PID_VERSION}/g;"
 
 if [ ! -f ${NIC_FW_FILE} ]; then
 	dd bs=4 count=1 if=/dev/zero of=${OUTPUT_DIR}/empty_nic_fw.bin
diff --git a/board/habanalabs/common/post-image-spi-arc_global.sh b/board/habanalabs/common/post-image-spi-arc_global.sh
index 38b6f7f..0f1e6d5 100755
--- a/board/habanalabs/common/post-image-spi-arc_global.sh
+++ b/board/habanalabs/common/post-image-spi-arc_global.sh
@@ -22,9 +22,9 @@
 PIF_ICCM_BINFILE=${PIF_ICCM_BINFILE:-pcie_pif-iccm.bin}
 PIF_DCCM_BINFILE=${PIF_DCCM_BINFILE:-pcie_pif-dccm.bin}
 
-ZEPHYR_PID_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-pid
-ZEPHYR_MGMT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-mgmt
-ZEPHYR_EVNT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/zephyr-custom/build-event
+PID_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/pid-custom
+MGMT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/mgmt-custom
+EVENT_DIR=${BR2_EXTERNAL}/output/${DEV_PREFIX}_arc/build/event-custom
 
 PRIME_IMAGE=0
 SECOND_IMAGE=1
diff --git a/board/habanalabs/gaudi2/post-image-agents.sh b/board/habanalabs/gaudi2/post-image-agents.sh
index 9226835..07266ff 100755
--- a/board/habanalabs/gaudi2/post-image-agents.sh
+++ b/board/habanalabs/gaudi2/post-image-agents.sh
@@ -25,7 +25,7 @@
 AGENT_ARCH="arm64"
 LOADABLES="security\", \"ppboot\", \"preboot\", \"img-ptr"
 
-HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/zephyr-custom/build-preboot/buildversion-habanalabs`
+HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/preboot-custom/build/buildversion-habanalabs`
 HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}_arc/build/baremetal-custom/buildversion-habanalabs | cut -d ' ' -f1`
 HL_FW_VERSION=`cut -d- -f2- ${BASE_DIR}/build/linux-custom/localversion-habanalabs`
 # Set hl-*-version = "hl-<device>-<HL_VERSION>[-rc]-fw-<fw ver>[-<git data>]"; inside FIT header
diff --git a/board/habanalabs/gaudi3/post-image-fit.sh b/board/habanalabs/gaudi3/post-image-fit.sh
index 0a0d69c..637c622 100644
--- a/board/habanalabs/gaudi3/post-image-fit.sh
+++ b/board/habanalabs/gaudi3/post-image-fit.sh
@@ -69,9 +69,9 @@
 	hl_extract_zephyr_elf_sections
 popd
 
-MGMT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-mgmt/zephyr/zephyr.stat
-PID_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-pid/zephyr/zephyr.stat
-EVENT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-event/zephyr/zephyr.stat
+MGMT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/mgmt-custom/build/zephyr/zephyr.stat
+PID_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/pid-custom/build/zephyr/zephyr.stat
+EVENT_STAT=${BR2_EXTERNAL}/output/${HL_DEV}/build/event-custom/build/zephyr/zephyr.stat
 
 # Extract the start address from stat file
 ENTRY_ADDRESS=$(cat $MGMT_STAT | grep "Entry point address" | head -1)
diff --git a/board/habanalabs/gaudi3_arc/post-image-agents.sh b/board/habanalabs/gaudi3_arc/post-image-agents.sh
index f4434dc..821b6ea 100644
--- a/board/habanalabs/gaudi3_arc/post-image-agents.sh
+++ b/board/habanalabs/gaudi3_arc/post-image-agents.sh
@@ -8,7 +8,7 @@
 source ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/create-fake-fuse-image.sh
 
 HL_COMPRESS=none
-HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-preboot/buildversion-habanalabs`
+HL_PREBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/preboot-custom/build/buildversion-habanalabs`
 HL_PPBOOT_VERSION_SECUREBOOT=`cat ${BR2_EXTERNAL}/output/${HL_DEV}/build/baremetal-custom/buildversion-habanalabs | cut -d ' ' -f1`
 HL_UBOOT_VERSION=""
 HL_FW_VERSION=""
@@ -22,7 +22,7 @@
 
 LOADABLES="ppboot\", \"preboot\", \"img-ptr\", \"pif-iccm\", \"pif-dccm\", \"preboot_d1"
 
-BIN=${BR2_EXTERNAL}/output/${HL_DEV}/build/zephyr-custom/build-agent-fw-loader/zephyr/zephyr*.bin
+BIN=${BR2_EXTERNAL}/output/${HL_DEV}/build/agent-fwl-custom/build/zephyr/zephyr*.bin
 GAUDI3_AGENT_BIN=$(ls ${BIN} 2>/dev/null)
 GAUDI3_AGENT_STAT=$(echo $GAUDI3_AGENT_BIN | sed 's/\.bin/.stat/')
 GAUDI3_SPEC_FW_LOADER_ITS=$(ls ${BR2_EXTERNAL}/board/habanalabs/${HL_DEV}/agent-fw_loader.its)
diff --git a/package/event/event.mk b/package/event/event.mk
index c3634fd..a3332ed 100644
--- a/package/event/event.mk
+++ b/package/event/event.mk
@@ -27,7 +27,7 @@
 
 EVENT_CONF_EXTRA_CFLAGS = \
 		-include $(EMBEDDED_ROOT)/specs/$(EVENT_DEVICE_NAME)/$(EVENT_DEVICE_NAME)_boot_info.h \
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)
 
 EVENT_CONF_ENV = CROSS_COMPILE=$(TARGET_CROSS)
@@ -46,7 +46,7 @@
 EVENT_CONF_OPTS += -DBUILD_VERSION="$(EVENT_APP_VERSION)"
 EVENT_CONF_OPTS += -DBUILD_MACHINE="$(shell hostname -s)"
 EVENT_CONF_OPTS += -DCMAKE_C_FLAGS="\
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(EVENT_CONF_CFLAGS)"
 
 EVENT_CONF_OPTS += -DDTS_EXTRA_CPPFLAGS="-I$(SPECS_EMBEDDED_DIR)"
diff --git a/package/mgmt/mgmt.mk b/package/mgmt/mgmt.mk
index 240eb3b..114c17c 100644
--- a/package/mgmt/mgmt.mk
+++ b/package/mgmt/mgmt.mk
@@ -27,7 +27,7 @@
 
 MGMT_CONF_EXTRA_CFLAGS =\
 		-include $(EMBEDDED_ROOT)/specs/$(MGMT_DEVICE_NAME)/$(MGMT_DEVICE_NAME)_boot_info.h \
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)
 
 MGMT_CONF_ENV = CROSS_COMPILE=$(TARGET_CROSS)
@@ -46,7 +46,7 @@
 MGMT_CONF_OPTS += -DBUILD_VERSION="$(MGMT_APP_VERSION)"
 MGMT_CONF_OPTS += -DBUILD_MACHINE="$(shell hostname -s)"
 MGMT_CONF_OPTS += -DCMAKE_C_FLAGS="\
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(MGMT_CONF_CFLAGS)"
 
 MGMT_CONF_OPTS += -DDTS_EXTRA_CPPFLAGS="-I$(SPECS_EMBEDDED_DIR)"
diff --git a/package/pid/pid.mk b/package/pid/pid.mk
index 103186a..8632e3d 100644
--- a/package/pid/pid.mk
+++ b/package/pid/pid.mk
@@ -11,7 +11,7 @@
 
 PID_CONF_EXTRA_CFLAGS =\
 		-include $(EMBEDDED_ROOT)/specs/$(PID_DEVICE_NAME)/$(PID_DEVICE_NAME)_boot_info.h \
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)
 
 ###################################################
@@ -42,7 +42,7 @@
 PID_CONF_OPTS += -DBUILD_VERSION="$(PID_APP_VERSION)"
 PID_CONF_OPTS += -DBUILD_MACHINE="$(shell hostname -s)"
 PID_CONF_OPTS += -DCMAKE_C_FLAGS="\
-		$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS) \
+		$(CONF_DEV_EXTRA_CFLAGS) \
 		$(PID_CONF_CFLAGS)"
 
 PID_CONF_OPTS += -DDTS_EXTRA_CPPFLAGS="-I$(SPECS_EMBEDDED_DIR)"
diff --git a/package/version-hooks/version-hooks.mk b/package/version-hooks/version-hooks.mk
index fcb627a..c216d99 100644
--- a/package/version-hooks/version-hooks.mk
+++ b/package/version-hooks/version-hooks.mk
@@ -111,8 +111,6 @@
 BAREMETAL_POST_RSYNC_HOOKS += HL_VERSION_GEN
 ARCBOOT_POST_RSYNC_HOOKS += HL_VERSION_GEN
 BTL_POST_RSYNC_HOOKS += HL_BTL_VERSION_GEN
-ZEPHYR_POST_RSYNC_HOOKS += HL_VERSION_GEN
-ZEPHYR_POST_RSYNC_HOOKS += ZEPHYR_SUBPACKAGES_VERSION_GEN
 HL_FW_LOADER_POST_RSYNC_HOOKS += HL_VERSION_GEN
 HLS2_BMC_POST_RSYNC_HOOKS += HL_VERSION_GEN
 HLS3_BMC_POST_RSYNC_HOOKS += HL_VERSION_GEN
diff --git a/package/zephyr/fs1.mk b/package/zephyr/fs1.mk
index d0968cb..fbba9cf 100644
--- a/package/zephyr/fs1.mk
+++ b/package/zephyr/fs1.mk
@@ -3,7 +3,13 @@
 #adding additional check for fs1 to avoid
 #including SPECS_H15_DIR for other platforms
 
-ifeq ($(DEV_PREFIX),fs1)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS =  -I$(SPECS_H15_DIR)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(SPECS_H15_DIR)/fs1/asic_reg
-endif
+FS1_SPECS_EMBEDDED_DIR = $(SPECS_EMBEDDED_DIR)/fs
+
+FS1_CONF_DEV_EXTRA_CFLAGS =\
+	-I$(FS1_SPECS_EMBEDDED_DIR) \
+	-I$(FS1_SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+	-I$(SPECS_ROOT) \
+	-I$(SPECS_EXT_ROOT) \
+	-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg \
+	-I$(SPECS_H15_DIR) \
+	-I$(SPECS_H15_DIR)/$(DEV_PREFIX)/asic_reg
diff --git a/package/zephyr/gaudi2.mk b/package/zephyr/gaudi2.mk
new file mode 100644
index 0000000..79f42a3
--- /dev/null
+++ b/package/zephyr/gaudi2.mk
@@ -0,0 +1,13 @@
+#Buildroot is including all the makefiles for
+#the compilation of zephyr appliation Hence
+#adding additional check for gaudi3 to avoid
+#including SPECS_H9_DIR for other platforms
+
+GAUDI2_SPECS_EMBEDDED_DIR=$(SPECS_EMBEDDED_DIR)
+
+GAUDI2_CONF_DEV_EXTRA_CFLAGS =\
+	-I$(GAUDI2_SPECS_EMBEDDED_DIR) \
+	-I$(GAUDI2_SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+	-I$(SPECS_ROOT) \
+	-I$(SPECS_EXT_ROOT) \
+	-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg
diff --git a/package/zephyr/gaudi3.mk b/package/zephyr/gaudi3.mk
index 194f786..8c4fd02 100644
--- a/package/zephyr/gaudi3.mk
+++ b/package/zephyr/gaudi3.mk
@@ -3,7 +3,13 @@
 #adding additional check for gaudi3 to avoid
 #including SPECS_H9_DIR for other platforms
 
-ifeq ($(DEV_PREFIX),gaudi3)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS =  -I$(SPECS_H9_DIR)
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -I$(SPECS_H9_DIR)/gaudi3/asic_reg
-endif
+GAUDI3_SPECS_EMBEDDED_DIR=$(SPECS_EMBEDDED_DIR)
+
+GAUDI3_CONF_DEV_EXTRA_CFLAGS =\
+	-I$(GAUDI3_SPECS_EMBEDDED_DIR) \
+	-I$(GAUDI3_SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+	-I$(SPECS_ROOT) \
+	-I$(SPECS_EXT_ROOT) \
+	-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg \
+	-I$(SPECS_H9_DIR) \
+	-I$(SPECS_H9_DIR)/$(DEV_PREFIX)/asic_reg
diff --git a/package/zephyr/zephyr.mk b/package/zephyr/zephyr.mk
index 583bb70..04c07c8 100644
--- a/package/zephyr/zephyr.mk
+++ b/package/zephyr/zephyr.mk
@@ -2,284 +2,99 @@
 ZEPHYR_SITE = zephyr
 ZEPHYR_LICENSE = unclear
 
-ifeq ($(DEV_PREFIX),fs1)
-SPECS_EMB_PROJ_DIR = $(SPECS_EMBEDDED_DIR)/fs
-else
-SPECS_EMB_PROJ_DIR = $(SPECS_EMBEDDED_DIR)
-endif
-
 ZEPHYR_DEPENDENCIES = host-cmake host-lzop host-lz4
 
-ZEPHYR_CONF_ENV = CROSS_COMPILE=$(TARGET_CROSS)
-ZEPHYR_CONF_ENV += ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
-ZEPHYR_CONF_ENV += ZEPHYR_BASE=$(ZEPHYR_BUILDDIR)
+ZEPHYR_SRCDIR=$(EMBEDDED_DIR)/zephyr
+ZEPHYR_BUILD_DIR ?= $(@D)/zephyr
 
-ZEPHYR_CONF_OPTS = -UCMAKE_TOOLCHAIN_FILE
-ZEPHYR_CONF_OPTS += -DUSER_CACHE_DIR=$(@D)/.cache
-ZEPHYR_CONF_OPTS += -DCMAKE_EXE_LINKER_FLAGS_INIT=-Wl,-e__start
-ZEPHYR_CONF_OPTS += -DBUILD_VERSION="$(shell cut -d- -f2- $(ZEPHYR_BUILDDIR)/localversion-habanalabs)"
-ZEPHYR_CONF_OPTS += -DBUILD_MACHINE="$(shell hostname -s)"
-ZEPHYR_CONF_OPTS += -DCMAKE_C_FLAGS="-I$(SPECS_EMB_PROJ_DIR) -I$(SPECS_EMB_PROJ_DIR)/$(DEV_PREFIX) -I$(SPECS_EMB_PROJ_DIR)/common -I$(SPECS_ROOT) -I$(SPECS_EXT_ROOT) -I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg"
-ZEPHYR_CONF_OPTS += -DDTS_EXTRA_CPPFLAGS="-I$(SPECS_EMB_PROJ_DIR)"
-
-ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DCONFIG_HL_STACK_OVERFLOW_REDUCE_SIZE=$(STACK_REDUCTION_SIZE)
-
-ifeq ($(HL_TARGET),asic)
-	ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_ASIC
-else ifeq ($(HL_TARGET),pldm)
-	ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_PLDM
-	ifeq ($(PLDM_SKIP_UBOOT),1)
-		ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_PLDM_REGRESSION
-	endif
-else ifeq ($(HL_TARGET),onix)
-	ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_ONIX
-endif
-
+BUILDDIR = $(@D)
 DEVICE_NAME = $(firstword $(subst _, ,$(HL_DEV)))
 
-# zephyr uses west tool to perform repo management, installation
-# of the tools used for zephyr compilation. For our zephyr
-# compilation, we do not make use of west tool.
-# If west is installed on build machine, it looks for configuration
-# file and throws error if it doesn't find one. To solve this issue,
-# create dummy configuration file in .west folder relative to zephyr
-# build directory.
-ZEPHYR_WEST_SRC_FILE_PATH = $(BR2_EXTERNAL_HABANALABS_PATH)/package/zephyr
-ZEPHYR_WEST_DEST_DIR_PATH = $(BR2_EXTERNAL_HABANALABS_PATH)/output/$(HL_DEV)/build/.west
+COMMON_DRV_CONF_CFLAGS =\
+		-I$(BUILDDIR) \
+		-I$(BUILDDIR)/include \
+		-I$(BUILDDIR)/include/drivers/misc/hl_npu \
+		-I$(BUILDDIR)/include/drivers/misc/hl_npu/$(HL_DEV) \
+		-I$(BUILDDIR)/include/drivers/misc/hl_npu/common \
+		-I$(BUILDDIR)/drivers/misc/hl_npu \
+		-I$(BUILDDIR)/drivers/misc/hl_npu/$(HL_DEV) \
+		-I$(BUILDDIR)/drivers/misc/hl_npu/common \
+		-I$(BUILDDIR)/include/generated \
+		-I$(BUILDDIR)/include/generated/syscalls
 
-ZEPHYR_LOCAL_VERSION = $$(sed "s/^.*fw-\\([0-9]\\+\\.[0-9]\\+\\.[0-9]\\+\\).*$$/\\1/" $(ZEPHYR_BUILDDIR)/localversion-habanalabs)
+COMMON_CONF_CFLAGS =
 
-ZEPHYR_CONFIGURE_CMDS = $(INSTALL) -D -m 0755 $(ZEPHYR_WEST_SRC_FILE_PATH)/west.config $(ZEPHYR_WEST_DEST_DIR_PATH)/config ; cd $(ZEPHYR_BUILDDIR) ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#ifndef _HL_VERSION_" > include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define _HL_VERSION_" >> include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_MAJOR $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f1)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_MINOR $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f2)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#define HL_VERSION_PATCH $$(echo $(ZEPHYR_LOCAL_VERSION) | cut -d. -f3)" >> $(ZEPHYR_BUILDDIR)/include/hl_version.h ;
-ZEPHYR_CONFIGURE_CMDS += echo "\#endif" >> include/hl_version.h
-ZEPHYR_BUILD_CMDS = true
-ZEPHYR_INSTALL_TARGET_CMDS = true
+ifeq ($($(call UPPERCASE,$(DEV_PREFIX))_SPECS_EMBEDDED_DIR),)
+DEV_SPECS_EMBEDDED_DIR = $($(call UPPERCASE,$(DEV_PREFIX))_SPECS_EMBEDDED_DIR)
+else
+DEV_SPECS_EMBEDDED_DIR = $(SPECS_EMBEDDED_DIR)
+endif
 
-ZEPHYR_COMPONENTS_LIST =
+CONF_DEV_EXTRA_CFLAGS=$($(call UPPERCASE,$(DEV_PREFIX))_CONF_DEV_EXTRA_CFLAGS)
+COMMON_CONF_EXTRA_CFLAGS =\
+		-include $(EMBEDDED_ROOT)/specs/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h \
+		$(CONF_DEV_EXTRA_CFLAGS)
 
-# Arguments:
-# - module name
-# - component dir
-# - component name
-define DEFINE_MODULE_ENVVARS
-	$(eval MODULE_NAME=$(strip $1)) \
-	$(eval MODULE_NAME_UPPER_CASE=$(shell echo $(MODULE_NAME) | tr a-z A-Z)) \
-	$(eval MODULE_NAME_UPPER_CASE=$(strip $(MODULE_NAME_UPPER_CASE))) \
-	$(eval COMPONENT_NAME_UPPER_CASE=$(shell echo $3 | tr a-z A-Z)) \
-	$(eval COMPONENT_NAME_UPPER_CASE=$(strip $(COMPONENT_NAME_UPPER_CASE))) \
-	\
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_SRCDIR=$(EMBEDDED_DIR)/zephyr-modules/$(MODULE_NAME)) \
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_DIR=$2/modules/$(MODULE_NAME)) \
-	$(eval $(COMPONENT_NAME_UPPER_CASE)_$(MODULE_NAME_UPPER_CASE)_NAME=$3_$(MODULE_NAME))
-endef
+COMMON_SPECS_CFLAGS = \
+		-I$(DEV_SPECS_EMBEDDED_DIR) \
+		-I$(DEV_SPECS_EMBEDDED_DIR)/$(DEV_PREFIX) \
+		-I$(SPECS_ROOT) \
+		-I$(SPECS_EXT_ROOT) \
+		-I$(SPECS_ROOT)/$(DEV_PREFIX)/asic_reg
 
-define CREATE_APP_VERSION
--DAPP_NAME=$1 \
--DAPP_VERSION=$(shell cut -d- -f2- $(ZEPHYR_BUILDDIR)/$2/localversion-habanalabs)
-endef
+# The STACK_REDUCTION_SIZE is set at each app makefile
+ZEPHYR_CONF_DEV_EXTRA_CFLAGS += \
+		-DCONFIG_HL_STACK_OVERFLOW_REDUCE_SIZE=$(STACK_REDUCTION_SIZE)
+
+ifeq ($(HL_TARGET),pldm)
+ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_PLDM
+ifeq ($(PLDM_SKIP_UBOOT),1)
+ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_PLDM_REGRESSION
+endif
+endif
+
+ifeq ($(HL_TARGET),onix)
+ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_ONIX
+endif
+
+ifeq ($(HL_TARGET),asic)
+ZEPHYR_CONF_DEV_EXTRA_CFLAGS += -DTARGET_IS_ASIC
+endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_PID
-ifdef BR2_arcle
-ifneq ($(STACK_REDUCTION_SIZE),0)
-ZEPHYR_CONF_OPTS += -DOVERLAY_CONFIG=$(BR2_EXTERNAL_HABANALABS_PATH)/configs/zephyr_arc_pid_stack_defconfig
-endif
-endif
-ZEPHYR_COMPONENTS_LIST += pid
-ZEPHYR_PID_DIR = build-pid
-ZEPHYR_PID_NAME = zephyr_arc_pid
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_PID_DIR)
-ZEPHYR_CONF_OPTS_PID += -DBOARD=hl_$(DEVICE_NAME)_pid_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_PID += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_PID_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcpid
-ZEPHYR_CONF_OPTS_PID += $(ZEPHYR_PID_SRCDIR)
-PWR_BRK_DUP_ID = $(shell grep PWRBRK_GROUP_ID $(SPECS_EMB_PROJ_DIR)/$(DEVICE_NAME)/$(DEVICE_NAME)_mem_addr_dt.h | grep -P -o "\d+"|head -1)
-ZEPHYR_CONF_OPTS += -DPWRBRK_GROUP_ID=$(PWR_BRK_DUP_ID)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,arcpid,$(ZEPHYR_PID_DIR)) $(ZEPHYR_CONF_OPTS_PID) -B $(ZEPHYR_PID_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_PID_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PID_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_pid.elf
-ZEPHYR_PID_MODULES = mbedtls boot_info eeprom
-$(foreach module,$(ZEPHYR_PID_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_PID_DIR),$(ZEPHYR_PID_NAME)))
+HL_ZEPHYR_PACKAGES += pid
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_MGMT
-ifdef BR2_arcle
-ifneq ($(STACK_REDUCTION_SIZE),0)
-ZEPHYR_CONF_OPTS += -DOVERLAY_CONFIG=$(BR2_EXTERNAL_HABANALABS_PATH)/configs/zephyr_arc_mgmt_stack_defconfig
-endif
-endif
-ZEPHYR_COMPONENTS_LIST += mgmt
-ZEPHYR_MGMT_DIR = build-mgmt
-ZEPHYR_MGMT_NAME = zephyr_arc_mgmt
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_MGMT_DIR)
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -include $(SPECS_EMB_PROJ_DIR)/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_MGMT += -DBOARD=hl_$(DEVICE_NAME)_mgmt_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_MGMT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_MGMT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcmgmt
-ZEPHYR_CONF_OPTS_MGMT += $(ZEPHYR_MGMT_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,arcmgmt,$(ZEPHYR_MGMT_DIR)) $(ZEPHYR_CONF_OPTS_MGMT) -B $(ZEPHYR_MGMT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_MGMT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_MGMT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_mgmt.elf
-ZEPHYR_MGMT_MODULES = eeprom mbedtls boot_info i2c-indirect soc_init hl_shell mctp pldm
-$(foreach module,$(ZEPHYR_MGMT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_MGMT_DIR),$(ZEPHYR_MGMT_NAME)))
+HL_ZEPHYR_PACKAGES += mgmt
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_EVENT
-ifdef BR2_arcle
-ifneq ($(STACK_REDUCTION_SIZE),0)
-ZEPHYR_CONF_OPTS += -DOVERLAY_CONFIG=$(BR2_EXTERNAL_HABANALABS_PATH)/configs/zephyr_arc_event_stack_defconfig
-endif
-endif
-ZEPHYR_COMPONENTS_LIST += event
-ZEPHYR_EVENT_DIR = build-event
-ZEPHYR_EVENT_NAME = zephyr_arc_event
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_EVENT_DIR)
-ZEPHYR_CONF_OPTS_EVENT += -DBOARD=hl_$(DEVICE_NAME)_event_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_EVENT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_EVENT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/arcevent
-ZEPHYR_CONF_OPTS_EVENT += $(ZEPHYR_EVENT_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(ZEPHYR_CONF_OPTS_EVENT) -B $(ZEPHYR_EVENT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_EVENT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_EVENT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_event.elf
-ZEPHYR_EVENT_MODULES = boot_info mbedtls
-$(foreach module,$(ZEPHYR_EVENT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_EVENT_DIR),$(ZEPHYR_EVENT_NAME)))
+HL_ZEPHYR_PACKAGES += event
 endif
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_PREBOOT
-ifdef BR2_arcle
-ifneq ($(STACK_REDUCTION_SIZE),0)
-ZEPHYR_CONF_OPTS += -DOVERLAY_CONFIG=$(BR2_EXTERNAL_HABANALABS_PATH)/configs/zephyr_arc_preboot_stack_defconfig
-endif
-endif
-ZEPHYR_COMPONENTS_LIST += preboot
-ZEPHYR_PREBOOT_DIR = build-preboot
-ZEPHYR_PREBOOT_NAME = zephyr_preboot
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_CONF_OPTS_PREBOOT += -DBOARD=hl_$(DEVICE_NAME)_preboot_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_PREBOOT += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_CONF_OPTS_PREBOOT += -DZEPHYR_BUILD_DIR=$(ZEPHYR_BUILDDIR)
-ZEPHYR_CONF_OPTS_PREBOOT += -DTARGET=$(HL_TARGET)
-ZEPHYR_PREBOOT_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/preboot
-ZEPHYR_CONF_OPTS_PREBOOT += $(ZEPHYR_PREBOOT_SRCDIR)
-ifeq ($(PREBOOT_RUN_FROM_ICCM),true)
-ZEPHYR_CONF_OPTS_PREBOOT += -DCONFIG_PREBOOT_RUN_FROM_ICCM=y
-endif #PREBOOT_RUN_FROM_ICCM
-#TODO - TARGET_IS_ASIC: Need to check why on ZEPHYR_CONF_OPTS its not working as expected
-ifeq ($(HL_TARGET),asic)
-ZEPHYR_CONF_OPTS_PREBOOT += -DEXTRA_CFLAGS="-include $(SPECS_EMB_PROJ_DIR)/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-endif
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,preboot,$(ZEPHYR_PREBOOT_DIR)) $(ZEPHYR_CONF_OPTS_PREBOOT) -B $(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_PREBOOT_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/preboot.bin
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_preboot.elf
-ZEPHYR_PREBOOT_MODULES = fdt boot_info mbedtls eeprom soc_init hl_shell mctp pldm
-$(foreach module,$(ZEPHYR_PREBOOT_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_PREBOOT_DIR),$(ZEPHYR_PREBOOT_NAME)))
+HL_ZEPHYR_PACKAGES += preboot
 endif #BR2_PACKAGE_ZEPHYR_ARC_PREBOOT
 
 ifdef BR2_PACKAGE_ZEPHYR_ARC_PREBOOT_D1
-ZEPHYR_COMPONENTS_LIST += preboot_d1
-ZEPHYR_PREBOOT_D1_DIR = build-preboot-d1
-ZEPHYR_PREBOOT_D1_NAME = zephyr_preboot_d1
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_PREBOOT_D1_DIR)
-ZEPHYR_CONF_OPTS_PREBOOT_D1 += -DBOARD=hl_$(DEVICE_NAME)_prebootd1_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_PREBOOT_D1 += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_CONF_OPTS_PREBOOT_D1 += -DZEPHYR_BUILD_DIR=$(ZEPHYR_BUILDDIR)
-ZEPHYR_CONF_OPTS_PREBOOT_D1 += -DTARGET=$(HL_TARGET)
-ZEPHYR_PREBOOT_D1_SRCDIR =  $(EMBEDDED_DIR)/zephyr-apps/preboot/app/preboot_d1
-ZEPHYR_CONF_OPTS_PREBOOT_D1 += $(ZEPHYR_PREBOOT_D1_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,preboot,$(ZEPHYR_PREBOOT_D1_DIR)) $(ZEPHYR_CONF_OPTS_PREBOOT_D1) -B $(ZEPHYR_PREBOOT_D1_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_PREBOOT_D1_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_D1_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/preboot_d1.bin
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_PREBOOT_D1_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_preboot_d1.elf
-endif #BR2_PACKAGE_ZEPHYR_ARC_PREBOOT_D1
-
-ifdef BR2_PACKAGE_ZEPHYR_ARC_DSM
-ZEPHYR_COMPONENTS_LIST += dsm
-ZEPHYR_DSM_DIR = build-dsm
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_DSM_DIR)
-ZEPHYR_CONF_OPTS_DSM += -DBOARD=hl_$(DEVICE_NAME)_dsm_$(HL_TARGET)
-ZEPHYR_CONF_OPTS_DSM += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_DSM_SRCDIR = $(EMBEDDED_DIR)/zephyr-apps/pcie-ss/dsm
-ZEPHYR_CONF_OPTS_DSM += $(ZEPHYR_DSM_SRCDIR)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(ZEPHYR_CONF_OPTS_DSM) -B $(ZEPHYR_DSM_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_DSM_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_DSM_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_dsm.elf
+HL_ZEPHYR_PACKAGES += preboot-d1
 endif
-
 
 ifdef BR2_PACKAGE_ZEPHYR_AGENT_FW_LOADER
-ZEPHYR_COMPONENTS_LIST += fw_loader
-ZEPHYR_FW_LOADER_DIR = build-agent-fw-loader
-ZEPHYR_FW_LOADER_NAME = zephyr_agent_fw_loader
-ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_FW_LOADER_DIR)
-ifdef BR2_arcle
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-Werror -Wno-misleading-indentation -include $(SPECS_EMB_PROJ_DIR)/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DBOARD=hl_$(DEVICE_NAME)_agent_$(HL_TARGET)
-endif
-ifdef BR2_aarch64
-ZEPHYR_CONF_OPTS += -DEXTRA_CFLAGS="-mstrict-align -Werror -Wno-misleading-indentation -include $(SPECS_EMB_PROJ_DIR)/$(DEVICE_NAME)/$(DEVICE_NAME)_boot_info.h $(ZEPHYR_CONF_DEV_EXTRA_CFLAGS)"
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DBOARD=hl_$(DEVICE_NAME)_$(HL_TARGET)
-endif
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DDEVICE=$(DEVICE_NAME)
-ZEPHYR_FW_LOADER_SRCDIR += $(EMBEDDED_DIR)/zephyr-plugins/fw-loader
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += $(ZEPHYR_FW_LOADER_SRCDIR)
-ZEPHYR_CONF_OPTS_AGENT_FW_LOADER += -DFAKE_SECURITY=$(FAKE_SECURITY)
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,fw-loader,$(ZEPHYR_FW_LOADER_DIR)) $(ZEPHYR_CONF_OPTS_AGENT_FW_LOADER) -B $(ZEPHYR_FW_LOADER_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_FW_LOADER_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_FW_LOADER_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/zephyr_agent_fw_loader.bin
-# The stat file is used for extracting the binary's load address and start address
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_FW_LOADER_DIR)/zephyr/zephyr.stat $(BINARIES_DIR)/zephyr_agent_fw_loader.stat
-ZEPHYR_FW_LOADER_MODULES = boot_info eeprom fdt i2c-indirect lz4 mbedtls hl_shell soc_init
-$(foreach module,$(ZEPHYR_FW_LOADER_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_FW_LOADER_DIR),$(ZEPHYR_FW_LOADER_NAME)))
+HL_ZEPHYR_PACKAGES += agent-fwl
 endif
 
-ifdef BR2_arcle
-ZEPHYR_MARGIN_TOOL_AGENT_BOARD = hl_$(DEVICE_NAME)_agent_$(HL_TARGET)
-else
-ZEPHYR_MARGIN_TOOL_AGENT_BOARD = hl_$(DEVICE_NAME)_$(HL_TARGET)
+ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_HBM
+HL_ZEPHYR_PACKAGES += agent-hbm
 endif
 
-# first argument is agent name, second argument is additional configuration opts
-define CREATE_ZEPHYR_MARGIN_TOOL_AGENT
-$(eval AGENT_LOWER_CASE = $(strip $1)) \
-$(eval ZEPHYR_COMPONENTS_LIST += $(AGENT_LOWER_CASE)) \
-$(eval AGENT_UPPER_CASE = $(shell echo $(AGENT_LOWER_CASE) | tr a-z A-Z)) \
-$(eval AGENT_UPPER_CASE = $(strip $(AGENT_UPPER_CASE))) \
-$(eval ZEPHYR_AGENT_DIR = build-agent-$1) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_DIR = build-agent-$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_NAME = zephyr_agent_$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_CONF_DEFAULT ?= $(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT = -DBOARD=$(ZEPHYR_MARGIN_TOOL_AGENT_BOARD)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += -DDEVICE=$(DEVICE_NAME)) \
-$(eval ZEPHYR_$(AGENT_UPPER_CASE)_SRCDIR = $(EMBEDDED_DIR)/zephyr-plugins/margin-tools/$(AGENT_LOWER_CASE)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += $(ZEPHYR_$(AGENT_UPPER_CASE)_SRCDIR)) \
-$(eval ZEPHYR_CONF_OPTS_AGENT += $2) \
-$(eval ZEPHYR_CONFIGURE_CMDS += && $$(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $$(ZEPHYR_CONF_OPTS) $(ZEPHYR_CONF_OPTS_AGENT) -B $(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $$(@D)/$(ZEPHYR_AGENT_DIR)) \
-$(eval ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $$(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.bin $(BINARIES_DIR)/zephyr_agent_$1.bin) \
-$(eval ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $$(@D)/$(ZEPHYR_AGENT_DIR)/zephyr/zephyr.stat $(BINARIES_DIR)/zephyr_agent_$1.stat) \
-$(foreach module,$(ZEPHYR_$(AGENT_UPPER_CASE)_MODULES),$(call DEFINE_MODULE_ENVVARS,$(module),$(ZEPHYR_$(AGENT_UPPER_CASE)_DIR), $(ZEPHYR_$(AGENT_UPPER_CASE)_NAME)))
-endef
-
-ifdef BR2_PACKAGE_ZEPHYR_AGENT_HBM
-ZEPHYR_HBM_MODULES = eeprom mbedtls boot_info i2c-indirect soc_init hl_shell
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,hbm, $(HBM_EXTRA_CONF))
-ZEPHYR_CONFIGURE_CMDS += $(call CREATE_APP_VERSION,hbm,build-agent-hbm)
+ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_SERDES
+HL_ZEPHYR_PACKAGES += agent-serdes
 endif
 
-ifdef BR2_PACKAGE_ZEPHYR_AGENT_SERDES
-ZEPHYR_SERDES_MODULES = boot_info fdt i2c-indirect mbedtls soc_init hl_shell
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,serdes, $(SERDES_EXTRA_CONF))
-ZEPHYR_CONFIGURE_CMDS += $(call CREATE_APP_VERSION,serdes,build-agent-serdes)
-endif
-
-ifdef BR2_PACKAGE_ZEPHYR_AGENT_PCIE
-ZEPHYR_PCIE_MODULES = boot_info i2c-indirect mbedtls soc_init hl_shell
-$(call CREATE_ZEPHYR_MARGIN_TOOL_AGENT,pcie, $(PCIE_EXTRA_CONF))
-ZEPHYR_CONFIGURE_CMDS += $(call CREATE_APP_VERSION,pcie,build-agent-pcie)
+ifdef BR2_PACKAGE_ZEPHYR_ARM_AGENT_PCIE
+HL_ZEPHYR_PACKAGES += agent-pcie
 endif
 
 ifeq ($(ZEPHYR_UNIT_TEST),1)
@@ -293,41 +108,11 @@
 else
 ZEPHYR_CONF_OPTS_UTEST += $(EMBEDDED_DIR)/zephyr/samples/hello_world
 endif
-ZEPHYR_CONFIGURE_CMDS += && $(ZEPHYR_CONF_ENV) $(HOST_DIR)/bin/cmake $(ZEPHYR_CONF_OPTS) $(call CREATE_APP_VERSION,test,$(ZEPHYR_UTEST_DIR)) $(ZEPHYR_CONF_OPTS_UTEST) -B $(ZEPHYR_UTEST_DIR)
-ZEPHYR_BUILD_CMDS += && $(TARGET_MAKE_ENV) $(MAKE) -C $(@D)/$(ZEPHYR_UTEST_DIR)
-ZEPHYR_INSTALL_TARGET_CMDS += && $(INSTALL) -D -m 0644 $(@D)/$(ZEPHYR_UTEST_DIR)/zephyr/zephyr.elf $(BINARIES_DIR)/zephyr_utest.elf
 endif
 
-# Arguments:
-# - module (mbedtls, eeprom, etc.)
-define ZEPHYR_COMPONENT_MODULE_VERSION_HOOK
-	$(eval MODULE_NAME_UPPER_CASE=$(shell echo $1 | tr a-z A-Z))
-	$(eval MODULE_NAME_UPPER_CASE=$(strip $(MODULE_NAME_UPPER_CASE)))
+ZEPHYR_DEPENDENCIES += $(HL_ZEPHYR_PACKAGES)
+ZEPHYR_POST_RSYNC_HOOKS += HL_VERSION_GEN
 
-	$(call HL_VERSION_GEN, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_SRCDIR}, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_DIR}, ${${COMPONENT_NAME_UPPER_CASE}_${MODULE_NAME_UPPER_CASE}_NAME})
-endef
+zephyr-dirclean: $(addsuffix -dirclean,$(HL_ZEPHYR_PACKAGES))
 
-# Arguments:
-# - component (zephyr apps and plugins)
-define ZEPHYR_COMPONENT_VERSION_HOOK
-	$(eval COMPONENT_UPPER_CASE=$(shell echo $1 | tr a-z A-Z))
-	$(eval COMPONENT_UPPER_CASE=$(strip $(COMPONENT_UPPER_CASE)))
-	$(eval COMPONENT_NAME=${ZEPHYR_${COMPONENT_UPPER_CASE}_NAME})
-	$(eval COMPONENT_NAME_UPPER_CASE=$(shell echo ${COMPONENT_NAME} | tr a-z A-Z))
-	$(eval COMPONENT_NAME_UPPER_CASE=$(strip $(COMPONENT_NAME_UPPER_CASE)))
-
-	$(call HL_VERSION_GEN, ${ZEPHYR_${COMPONENT_UPPER_CASE}_SRCDIR}, ${ZEPHYR_${COMPONENT_UPPER_CASE}_DIR}, ${COMPONENT_NAME})
-
-	$(foreach module,$(ZEPHYR_$(COMPONENT_UPPER_CASE)_MODULES),$(call ZEPHYR_COMPONENT_MODULE_VERSION_HOOK,$(module)))
-endef
-
-define ZEPHYR_SUBPACKAGES_VERSION_GEN
-	$(foreach component,$(ZEPHYR_COMPONENTS_LIST),$(call ZEPHYR_COMPONENT_VERSION_HOOK,${component}))
-endef
-
-zephyr-configure: $(ZEPHYR_DEPENDENCIES)
-zephyr-menuconfig zephyr-guiconfig: | zephyr-configure
-	$(TARGET_MAKE_ENV) $(MAKE) -C $(ZEPHYR_BUILDDIR)/$(ZEPHYR_CONF_DEFAULT) $(patsubst zephyr-%,%,$@)
-	rm -f $(ZEPHYR_BUILDDIR)/.stamp_{configured,built}
-
-$(eval $(cmake-package))
+$(eval $(generic-package))
diff --git a/scripts/make/gaudi2/Makefile b/scripts/make/gaudi2/Makefile
index 3a258ce..733532e 100644
--- a/scripts/make/gaudi2/Makefile
+++ b/scripts/make/gaudi2/Makefile
@@ -188,8 +188,8 @@
 .PHONY: gaudi2-debug-elfs-create
 gaudi2-debug-elfs-create:
 	@mkdir -p $(IMAGES_DIR)/elfs
-	@for elf in $$(ls $(BR2_EXTERNAL)/output/{gaudi2,gaudi2_arc}/build/zephyr-custom/*/zephyr/zephyr.elf); do \
-		cp $${elf} $(IMAGES_DIR)/elfs/`echo $${elf} | sed "s@.*/\(.*\)/zephyr/zephyr.elf@\1.elf@"`; \
+	@for elf in $$(ls $(BR2_EXTERNAL)/output/{gaudi2,gaudi2_arc}/build/*-custom/zephyr/zephyr.elf); do \
+		cp $${elf} $(IMAGES_DIR)/elfs/`echo $${elf} | sed "s@.*/\(.*\)-custom/zephyr/zephyr.elf@\1.elf@"`; \
 	done
 	@tar czf $(IMAGES_DIR)/${HL_DEV}-elfs.tgz -C $(IMAGES_DIR)/ elfs
 	@echo
diff --git a/scripts/make/gaudi3/Makefile b/scripts/make/gaudi3/Makefile
index 251f3b3..5441eaf 100644
--- a/scripts/make/gaudi3/Makefile
+++ b/scripts/make/gaudi3/Makefile
@@ -7,8 +7,8 @@
 .PHONY: gaudi3-debug-elfs-create
 gaudi3-debug-elfs-create:
 	@mkdir -p $(IMAGES_DIR)/elfs
-	@for elf in $$(ls $(BR2_EXTERNAL)/output/gaudi3_arc/build/zephyr-custom/*/zephyr/zephyr.elf); do \
-		cp $${elf} $(IMAGES_DIR)/elfs/`echo $${elf} | sed "s@.*/\(.*\)/zephyr/zephyr.elf@\1.elf@"`; \
+	@for elf in $$(ls $(BR2_EXTERNAL)/output/gaudi3_arc/build/*-custom/zephyr/zephyr.elf); do \
+		cp $${elf} $(IMAGES_DIR)/elfs/`echo $${elf} | sed "s@.*/\(.*\)-custom/zephyr/zephyr.elf@\1.elf@"`; \
 	done
 	@tar czf $(IMAGES_DIR)/${HL_DEV}-elfs.tgz -C $(IMAGES_DIR)/ elfs
 	@echo
